### ISSUE TITLE
openfl.events.Event.target confusingly typed

### ISSUE THREAD

**Issue by ProPuke** [openfl/openfl#1321]
`openfl.events.Event.target` is currently typed as `Dynamic`, not `DisplayObject`.

This means if you go to modify any of its properties it will silently fail (due to accessors not being called), if you go to read you will only receive `null`, but if you `trace` it you will be informed that is indeed a valid display object.  This is counter intuitive.

For example, consider the following event handler:

```
public onMouseoverObject(e:MouseEvent):Void {
    trace('mouse over ${e.target}');
    e.target.scaleX = 2;
    e.target.scaleY = 2;
}
```

This will, for example, print `"mouse over [Sprite]"`, but then do nothing, while presenting no feedback to the user as to the nature of the mistake. Instead it currently needs to be:

```
public onMouseoverObject(e:MouseEvent):Void {
    var target:DisplayObject = e.target;
    trace('mouse over ${target}');
    target.scaleX = 2;
    target.scaleY = 2;
}
```

This will correctly modify the object, as `set_scaleX` etc will now be called.

Currently it seems very easy to accidentally catch people out here. The docs aren't clear on differentiating what are properties and what are variables, and thus what will and will-not work through a `Dynamic`. Haxe is otherwise a strictly-typed language. Lets embrace some of that safety!

**Comment by jgranick** [comment]
Strictly speaking, I think the `event.target` will be an `IEventDispatcher`, not necessarily a `DisplayObject`.

```haxe
var dispatcher = new EventDispatcher ();
dispatcher.addEventListener ("test", function (e) trace (e.target));
dispatcher.dispatchEvent (new Event ("test"));
```

If we type to `IEventDispatcher` (which might also be negotiable) then that doesn't really help. The Haxe C++ target used to work above without typing, but this was broken in Haxe some versions ago, for consistency with other targets. The official behavior for getters and setters is you must type the object properly.

We could consider using a generic internally, so MouseEvent is an Event<DisplayObject>, but that excludes the possibility of custom mouse events (dispatched from some other object) and breaks otherwise valid code when you try `new Event` instead of `new Event<T>`

I can't think of a better solution to this without changing the API

**Comment by ProPuke** [comment]
I'd say even using `IEventDispatcher` would be a better idea. You'd still have to cast it as you do currently, but at least there'd be type-safety errors if you don't. `Dynamic` currently gives no feedback if you do the wrong thing :S

**Comment by jgranick** [comment]
Now that I have thought about it more, I agree.

In ActionScript 3, being dynamic allows you to do things like:

```as3
var loader = new Loader ();
loader.contentLoaderInfo.addEventListener (Event.COMPLETE, function (e) {
   trace (e.currentTarget.content);
});
```

...this is now broken, unless you use a cast. However, this can easily be rewritten as

```haxe
var loader = new Loader ();
loader.contentLoaderInfo.addEventListener (Event.COMPLETE, function (e) {
   trace (loader.content);
});
```

This is less convenient, but is better for performance, and due to the "contract" between `EventDispatcher` and `Event`, I'm quite sure that `target` and `currentTarget` will be an `IEventDispatcher`

Even though the dynamic is more convenient in AS3, breaking on Haxe getters and setters is not fun.

I also found (as it would happen) a core event dispatcher that relied upon the `target` being `Dynamic`, now that it is cast properly, there may be a significant performance boost on C++ there.

Lets see what happens, but otherwise, this should now give an error if it is not cast

**Comment by rmsmith** [comment]
Hello, this broke the build of the `swf` library:

```
../format/swf/tags/TagDefineBitsJPEG3.hx:72: characters 22-49 : Cast type must be a class or an enum
../format/swf/tags/TagDefineBits.hx:71: characters 22-41 : openfl.events.IEventDispatcher has no field loader
```

**Comment by vroad** [comment]
This broke the build of Starling v2.0 as well. Should I fix this by changing starling code?

**Comment by ProPuke** [comment]
Projects using `openfl.events.Event.target` will have to change their code, yes.
`openfl.events.Event.target` is now typed, so you will need to cast it to the required type in order to use it (or access an outer variable as @jgranick did in his example).

This is actually a good thing, as it's more performant in typed languages, and you'll get proper type-safe errors if you accidentally typo or try and access a property that doesn't exist.

**Comment by jgranick** [comment]
This is how I fixed it in 1.8

https://github.com/openfl/starling/commit/323d073bc2d354f5e032b7881390d4c3604da0f7

It's a nuisance (I know) but since Haxe getters and setters don't behave like a real field, I think that all things considered, this change is the best since it keeps things safe, and will lead to minor performance improvements

**Comment by sourceskyboxer** [comment]
Still OpenFL 5.1.2 has problems event.target.<targetmethod> like event.currentTarget.<targetmethod> too

I can not understand why does it happen again new version of OpenFL has problem :( 

Please fix OpenFL next release
